bin/taiyi-underwriter-monitor/src/main.rs

```diff
@@ -0,0 +1,279 @@
+                    })
+                    .collect::<Vec<_>>();
+                let first_elem = prices.first().expect("bug, should have at least one hash");
+                if preconf_type == 0 {
+                    // Type A
+                    assert!(
+                        prices.iter().all(|x| x == first_elem),
+                        "type a bug: all transactions should have same gas price"
+                    );
+                } else {
+                    // Type B
+                    assert!(prices.len() == 1, "type b bug: there shouldn't be more than one tx");
+                }
+                *first_elem
+            };
+
+            UnderwriterTradeRow::update_with_settlement(
+                preconf.uuid,
+                u128_to_big_decimal(realized_gas_price),
+                realized_blob_price.as_ref().map(|x| u128_to_big_decimal(*x)).ok(),
+                &db_conn,
+            )
+            .await?;
+        }
+
+        debug!("[Challenger Creator]: Processed block {:?}", header.number);
+    }
+
+    Ok(())
+}
+
+fn get_blob_gas_price(data: &Block) -> eyre::Result<u128> {
+    let excess_blob_gas =
+        data.header.excess_blob_gas.ok_or_eyre("no excess blob gas in block header")? as f64;
+    let base_fee_per_blob_gas = MIN_BASE_FEE_PER_BLOB_GAS as f64
+        * (excess_blob_gas / BLOB_BASE_FEE_UPDATE_FRACTION as f64).exp();
+    Ok(base_fee_per_blob_gas as u128)
+}
+
+async fn get_genesis_time(beacon_url: &str) -> eyre::Result<u64> {
+    // Read genesis timestamp from Beacon API (/eth/v1/beacon/genesis)
+    let beacon_genesis_response = match reqwest::Client::new()
+        .get(format!("{beacon_url}/eth/v1/beacon/genesis"))
+        .send()
+        .await
+    {
+        Ok(response) => response,
+        Err(e) => {
+            error!("Failed to get beacon genesis: {e}");
+            return Err(eyre::eyre!("Failed to get beacon genesis: {e}"));
+        }
+    };
+
+    let beacon_genesis_response = match beacon_genesis_response.json::<serde_json::Value>().await {
+        Ok(value) => value,
+        Err(e) => {
+            error!("Failed to parse beacon genesis response: {e}");
+            return Err(eyre::eyre!("Failed to parse beacon genesis response: {e}"));
+        }
+    };
+
+    let genesis_time =
+        beacon_genesis_response["data"]["genesis_time"].as_str().ok_or_else(|| {
+            let err = "Failed to get genesis time from response";
+            error!("{}", err);
+            eyre::eyre!("{}", err)
+        })?;
+
+    let genesis_timestamp = match u64::from_str(genesis_time) {
+        Ok(timestamp) => timestamp,
+        Err(e) => {
+            error!("Failed to parse genesis time: {}", e);
+            return Err(eyre::eyre!("Failed to parse genesis time: {}", e));
+        }
+    };
+    Ok(genesis_timestamp)
+}
+
+#[derive(Parser, Clone)]
+struct Opts {
+    #[clap(long)]
+    execution_client_ws_url: String,
+    #[clap(long)]
+    beacon_url: String,
+}
```

Comment:
```suggestion
    #[clap(long]
    pub network: String,
}
```

Consider adding this extra field for network. Valid inputs are: `mainnet`, `holesky`, etc